### DISCUSSION TITLE
Some issues about use DashBoard

### DISCUSSION THREAD

**Discussion by gmij** [microsoft/aspire#2775] [Q&A]
**1. How to add an external ResourceServer?**

In DashBoardClient, you can see that the retrieval of resources is obtained from the address ResourceServiceUrlVariableName. But there is no place to declare this variable. And this variable is currently not initialized in the environment variable.

**2. If DashBoard is registered as a container in Host and published as compose.yml, as follows:**

`    Var dashboard=builder AddContainer ("dashboard", "mcr. microsoft. com/dotnet/nightly/aspire dashboard");`
`    Dashboard WithEndpoint (18889, 4317, name: "OTEL");`
`    Dashboard WithEndpoint (18888, hostPort: 33456, scheme: "http", name: "webui");`
`    Dashboard AsHttp2Service(); `

At this point, other resources cannot be published to the dashboard, and can only be set in the environment parameters of the resources as follows:

  `
  ApiService WithEnvironment ("OTEL-EXPORTER-OTLP_ENDPoint",“ http://dashboard:18889/" );
  `

In this case, the resource and console log panels cannot be seen in the dashboard.

3. If in the dashboard of the second point, you try to configure DOTNET-RESOURCE-SERVICE-ENDPOINTURL in the environment variable, and at this point, you don't know who to point to. In theory, it should point to the dashboard itself. But after trying, it cannot be used normally.

------------------------------

1. 如何添加外部的ResourceServer？
    在DashBoardClient中，可以看到资源的获取，是从ResourceServiceUrlVariableName这个地址中获取的。 但是没有地方可以去声明这个变量。且这个变量，当前也并没有在环境变量中初始化的情况。
2. 如果把DashBoard做为容器注册在Host中，并发布成compose.yml，如下：
  ` csharp
  
      var dashboard = builder.AddContainer("dashboard", "mcr.microsoft.com/dotnet/nightly/aspire-dashboard");
      dashboard.WithEndpoint(18889, 4317, name: "OTEL");
      dashboard.WithEndpoint(18888, hostPort: 33456, scheme: "http", name: "webui");
      dashboard.AsHttp2Service();
  `

这时候，无法将其它的资源，发布到dashboard中，仅能在资源的环境参数中，设置如下：

  `csharp
     apiService.WithEnvironment("OTEL_EXPORTER_OTLP_ENDPOINT", "http://dashboard:18889/");
  
  `
这样的话，在dashboard中，是无法看到resource和console log面板的。

3. 如果在第二点的dashboard中，尝试在环境变量中配置 DOTNET_RESOURCE_SERVICE_ENDPOINT_URL， 这时候不知道要指向谁了。理论上，应该是指向dashboard自身。但是尝试后，无法正常使用。

**Comment by davidfowl** [comment]
> In DashBoardClient, you can see that the retrieval of resources is obtained from the address ResourceServiceUrlVariableName. But there is no place to declare this variable. And this variable is currently not initialized in the environment variable.

The only resource server that ships as part of the project is the AppHost. If you run the dashboard standalone, there is no resource server provided for you. 

> In this case, the resource and console log panels cannot be seen in the dashboard.

That's correct.

> If in the dashboard of the second point, you try to configure DOTNET-RESOURCE-SERVICE-ENDPOINTURL in the environment variable, and at this point, you don't know who to point to. In theory, it should point to the dashboard itself. But after trying, it cannot be used normally.

It shouldn't point to the dashboard itself. The dashboard isn't a resource server. 

I wrote more on this [here](https://github.com/dotnet/aspire/discussions/1439#discussioncomment-8737953) and https://github.com/prom3theu5/aspirational-manifests/issues/118

**Reply by gmij** [reply]
> The only resource server that ships as part of the project is the AppHost. If you run the dashboard standalone, there is no resource server provided for you.

Okay, how do we publish the Dashboard and the application together on the production line? DashBoard is very useful for lightweight businesses (it includes commonly used monitoring information such as log viewing, APM, performance metrics, etc.)

**Reply by davidfowl** [reply]
You can use it as an otlp server. If you want the resource view then more work needs to be done to build a resource server for your environment.

**Reply by davidfowl** [reply]
I’d suggest contributing ideas and code to the aspirate project!
